homeassistant/components/template/lock.py

```diff
@@ -130,24 +137,67 @@ def _update_state(self, result):
 
         self._state = None
 
+    @property
+    def code_format(self) -> str | None:
+        """Regex for code format or None if no code is required."""
+        return self._code_format
+
     @callback
     def _async_setup_templates(self) -> None:
         """Set up templates."""
         self.add_template_attribute(
             "_state", self._state_template, None, self._update_state
         )
+        if self._code_format_template:
+            self.add_template_attribute(
+                "_code_format_template",
+                self._code_format_template,
+                None,
+                self._update_code_format,
+            )
         super()._async_setup_templates()
 
+    @callback
+    def _update_code_format(self, render: str | TemplateError | None):
+        """Update code format from the template."""
+        if isinstance(render, TemplateError):
+            self._code_format = None
+            self._code_format_template_error = render
+        elif render in (None, "None", ""):
+            self._code_format = None
+            self._code_format_template_error = None
+        else:
+            self._code_format = render
+            self._code_format_template_error = None
+
     async def async_lock(self, **kwargs: Any) -> None:
         """Lock the device."""
+        self._raise_template_error_if_available()
+
         if self._optimistic:
             self._state = True
             self.async_write_ha_state()
-        await self.async_run_script(self._command_lock, context=self._context)
+
+        tpl_vars = {ATTR_CODE: kwargs.get(ATTR_CODE) if kwargs else None}
+
+        await self.async_run_script(
+            self._command_lock, run_variables=tpl_vars, context=self._context
+        )
 
     async def async_unlock(self, **kwargs: Any) -> None:
         """Unlock the device."""
+        self._raise_template_error_if_available()
+
         if self._optimistic:
             self._state = False
             self.async_write_ha_state()
-        await self.async_run_script(self._command_unlock, context=self._context)
+
+        tpl_vars = {ATTR_CODE: kwargs.get(ATTR_CODE) if kwargs else None}
+
+        await self.async_run_script(
+            self._command_unlock, run_variables=tpl_vars, context=self._context
+        )
+
+    def _raise_template_error_if_available(self):
+        if self._code_format_template_error is not None:
+            raise ServiceValidationError(self._code_format_template_error)
```

Comment:
Nice!
Please customize this so it's clear what template errored, use a translated message with placeholders for the name of the failing template and the error message in the _code_format_template_error

Reply:
Now:
![grafik](https://github.com/home-assistant/core/assets/57918757/ac9e97f5-314c-446c-8146-69891a6c94c8)